TraceView: Resource attributes links extension point

edvard-falkskar opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

7. If your PR content should be added to the What's New document for the next major or minor release, add the **add to what's new** label to your PR. Note that you should add this label to the main PR that introduces the feature; do not add this label to smaller PRs for the feature.

-->

**What is this feature?**

Introduces a new plugin extension point for trace view resource attributes that lets plugins turn resource attributes into clickable links. Plugins are passed a context containing information about the data source and the resource attributes in the span. The plugin link category prop is used to let trace view know which resource attribute to turn into a link.

[Example implementation of links in app o11y](https://github.com/grafana/app-observability-plugin/pull/1771/files).

[Corresponding work for Logs view](https://github.com/grafana/grafana/pull/105943)


![image](https://github.com/user-attachments/assets/c1faa89a-d16e-4080-9490-9dc8b3e40567)


**Why do we need this feature?**

Part of the initiative to better connect act 3 solutions. With this solution act 3 solutions can target resource attributes relevant to it's area. As an example App o11y can add links for `service.name` and `service.namespace` that turns these attributes into links that takes the user into the App o11y service overview. This helps end users that no longer has to copy paste values from one place to another.

**Who is this feature for?**

Plugin developers.

**Which issue(s) does this PR fix?**:

<!--

- Automatically closes linked issue when the Pull Request is merged.

Fixes: https://github.com/grafana/app-observability-plugin/issues/1606
-->

Fixes https://github.com/grafana/app-observability-plugin/issues/1606

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/contribute/release-notes/#how-to-determine-if-content-belongs-in-whats-new), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/contribute/release-notes/) doc.

edvard-falkskar committed:
TraceView: Resource attributes links extension point

edvard-falkskar committed:
Add data source info to context

edvard-falkskar committed:
Remove console log

ifrost commented on packages/grafana-data/src/types/pluginExtensions.ts:
```diff
@@ -192,6 +192,7 @@ export enum PluginExtensionPoints {
   UserProfileTab = 'grafana/user/profile/tab',
   TraceViewDetails = 'grafana/traceview/details',
   QueryEditorRowAdaptiveTelemetryV1 = 'grafana/query-editor-row/adaptivetelemetry/v1',
+  ResourceAttributes = 'grafana/resource-attributes',
```
Do we plan to use this extension outside of TraceView as well? And with other data sources? I'm wondering if it should be as broadly generic as `grafana/resource-attributes` or something like `grafana/traceview/resource-attributes` or `grafana/traces/resource-attributes` 🤔

edvard-falkskar replied:
We want to do the same thing for logs at least, but might not be a bad idea to be able so scope it anyway.

edvard-falkskar replied:
Changed it to be `grafana/traceview/resource-attributes`

edvard-falkskar committed:
Merge remote-tracking branch 'origin/main' into edvard/trace-view-resource-attributes-extension

edvard-falkskar committed:
Removing unused linkGetters and more

edvard-falkskar committed:
More tests

edvard-falkskar committed:
Fixing last todos

edvard-falkskar committed:
Merge remote-tracking branch 'origin/main' into edvard/trace-view-resource-attributes-extension

edvard-falkskar committed:
Fixing

edvard-falkskar committed:
Fixing

nadinevehling reviewed (commented):
Hey Edvard,
I already asked you this in Slack but like to post it here in context: Where do users go? If this is different per configuration a tooltip that explains general behavior can help users to know what to expect.

Is the A the newest logo of O11y?

I'd like to see if changing the style of the selectable links to the "blue link" color differentiates the selectable attributes from normal non-selectable attributes better.

edvard-falkskar commented:
Hey @nadinevehling

> Hey Edvard, I already asked you this in Slack but like to post it here in context: Where do users go? If this is different per configuration a tooltip that explains general behavior can help users to know what to expect.

Where the user goes is up to the plugin author who is adding a link to one of the attributes, I have added the tooltip like we talked about before (see final image of this comment).

> Is the A the newest logo of O11y?

It should be, it's the one from our [Storybook Icon component](https://developers.grafana.com/ui/latest/index.html?path=/story/docs-overview-icon--icons-overview).

![Screenshot 2025-05-20 at 08 56 44](https://github.com/user-attachments/assets/1f9571cf-d6fc-4411-a494-261478d4ffeb)

> I'd like to see if changing the style of the selectable links to the "blue link" color differentiates the selectable attributes from normal non-selectable attributes better.

Like this? I changed the underline to only be shown when hovering the link.

![Screenshot 2025-05-20 at 08 54 11](https://github.com/user-attachments/assets/9f4c8f13-c5df-4dd5-91d3-ecf34c1f574b)

edvard-falkskar committed:
Change link style

edvard-falkskar committed:
Merge remote-tracking branch 'origin/main' into edvard/trace-view-resource-attributes-extension

hugohaggmark reviewed (commented):
✅  approved from the plugins-platform

edvard-falkskar committed:
Rename extension context

edvard-falkskar committed:
Merge remote-tracking branch 'origin/main' into edvard/trace-view-resource-attributes-extension

edvard-falkskar committed:
Fix lint error

edvard-falkskar committed:
Merge remote-tracking branch 'origin/main' into edvard/trace-view-resource-attributes-extension

ifrost commented on packages/grafana-data/src/types/pluginExtensions.ts:
```diff
@@ -236,6 +237,15 @@ export type PluginExtensionDataSourceConfigContext<
 
+export type PluginExtensionResourceAttributesContext = {
+  // Key-value pairs of resource attributes, attribute name is the key
+  attributes: Record<string, string[]>;
```
This might be a more open question. With this code we're gonna pass raw attributes here coming from Tempo or Loki (and other data sources in the future). This means app-o11y will try to figure out the meaning of the attribute based on the name like it's now done in https://github.com/grafana/app-observability-plugin/pull/1771/files:

```
    .addLink(makeTraceViewAttributesLink('service.name', plugin))
    .addLink(makeTraceViewAttributesLink('service.namespace', plugin))
    .addLink(makeLogsViewAttributesLink('service_name', plugin))
    .addLink(makeLogsViewAttributesLink('service_namespace', plugin));
```

I was wondering what if we started using Open Telemetry as a common format instead. Tempo and Loki plugins (and others) would be responsible for figuring out which tag/attribute represents "service name" (or any other OTEL attributes in future) and app-o11y would not need to know about how to extract it. So for example the context for the app-o11y extension could look like:

```
import { Resource } from '@opentelemetry/resources';

export type PluginExtensionResourceAttributesContext = {
  resource: Resource;
  datasource: {
    type: string;
    uid: string;
  };
};
```

And the plugin code (Tempo, Loki, etc.) would need to transform it's own data into OTEL-compatible type, e.g.:

```
const resource = new Resource({
  'service.name': '...',
});
```

The app-o11y would not need to check specific attribute name (`service.name`', `service_name` and would use OTEL's `Resource` instead).

This way OTEL would become common format to be shared between various signals/plugins.
